sdk/include/opentelemetry/sdk/metrics/exemplar/histogram_exemplar_reservoir.h

```diff
@@ -0,0 +1,80 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#ifndef ENABLE_METRICS_PREVIEW
+#  include <memory>
+#  include <vector>
+#  include "opentelemetry/context/context.h"
+#  include "opentelemetry/nostd/shared_ptr.h"
+#  include "opentelemetry/sdk/common/attribute_utils.h"
+#  include "opentelemetry/sdk/metrics/data/exemplar_data.h"
+#  include "opentelemetry/sdk/metrics/exemplar/filter.h"
+#  include "opentelemetry/sdk/metrics/exemplar/fixed_size_exemplar_reservoir.h"
+#  include "opentelemetry/sdk/metrics/exemplar/reservoir.h"
+#  include "opentelemetry/sdk/metrics/exemplar/reservoir_cell_selector.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace metrics
+{
+
+class HistogramExemplarReservoir : public FixedSizeExemplarReservoir<ExemplarData>
+{
+
+public:
+  static std::shared_ptr<ReservoirCellSelector> GetHistogramCellSelector(
+      const std::vector<double> &boundaries = std::vector<double>{1.0, 2.0, 3.0, 4.0, 5.0})
+  {
+    return std::shared_ptr<ReservoirCellSelector>{new HistogramCellSelector(boundaries)};
+  }
+
+  HistogramExemplarReservoir(size_t size,
+                             std::shared_ptr<ReservoirCellSelector> reservoir_cell_selector,
+                             std::shared_ptr<ExemplarData> (ReservoirCell::*map_and_reset_cell)(
+                                 const common::OrderedAttributeMap &attributes))
+      : FixedSizeExemplarReservoir<ExemplarData>(size, reservoir_cell_selector, map_and_reset_cell)
+  {}
+
+  class HistogramCellSelector : public ReservoirCellSelector
+  {
+  public:
+    HistogramCellSelector(const std::vector<double> &boundaries) : boundaries_(boundaries) {}
+
+    int ReservoirCellIndexFor(const std::vector<ReservoirCell> &cells,
+                              long value,
+                              const MetricAttributes &attributes,
+                              const opentelemetry::context::Context &context) override
+    {
+      return ReservoirCellIndexFor(cells, (double)value, attributes, context);
+    }
+
+    int ReservoirCellIndexFor(const std::vector<ReservoirCell> &cells,
+                              double value,
+                              const MetricAttributes &attributes,
+                              const opentelemetry::context::Context &context) override
+    {
+      for (size_t i = 0; i < boundaries_.size(); ++i)
```

Comment:
nit - this reminds me of the possibility of using binary search for over-optimization. 
Just kidding, ignore it :)